mirgecom/simutil.py

```diff
@@ -401,37 +404,143 @@ def generate_and_distribute_mesh(comm, generate_mesh):
         The number of elements in the serial mesh
     """
-    from meshmode.distributed import (
-        MPIMeshDistributor,
-        get_partition_by_pymetis,
-    )
-    num_parts = comm.Get_size()
-    mesh_dist = MPIMeshDistributor(comm)
-    global_nelements = 0
+    from warnings import warn
+    warn(
+        "generate_and_distribute_mesh is deprecated and will go away Q4 2022. "
+        "Use distribute_mesh instead.", DeprecationWarning, stacklevel=2)
+    return distribute_mesh(comm, generate_mesh)
 
-    if mesh_dist.is_mananger_rank():
 
-        mesh = generate_mesh()
+def distribute_mesh(comm, get_mesh_data, partition_generator_func=None):
+    # FIXME: Out of date
+    """Distribute a mesh among all ranks in *comm*.
 
-        global_nelements = mesh.nelements
+    Retrieve the global mesh data with the user-supplied function *get_mesh_data*,
+    partition the mesh, and distribute it to every rank in the provided MPI
+    communicator *comm*.
 
-        part_per_element = get_partition_by_pymetis(mesh, num_parts)
-        local_mesh = mesh_dist.send_mesh_parts(mesh, part_per_element, num_parts)
-        del mesh
+    .. note::
+        This is a collective routine and must be called by all MPI ranks.
 
-    else:
-        local_mesh = mesh_dist.receive_mesh_part()
+    Parameters
+    ----------
+    comm:
+        MPI communicator over which to partition the mesh
+    get_mesh_data:
+        Callable of zero arguments returning *mesh* or *(mesh, volume_to_elements)*,
+        where *mesh* is a :class:`meshmode.mesh.Mesh` and *volume_to_elements* is a
+        :class:`dict` mapping volume tags to :class:`numpy.ndarray`s of element
+        numbers.
 
-    return local_mesh, global_nelements
+    Returns
+    -------
+    local_mesh_data : :class:`meshmode.mesh.Mesh` or :class:`dict`
+        If *get_mesh_data* returns only a mesh, *local_mesh_data* is the local mesh.
+        If *get_mesh_data* also returns *volume_to_elements*, *local_mesh_data* will
+        be a :class:`dict` mapping volume tags to corresponding local meshes.
+    global_nelements : :class:`int`
+        The number of elements in the global mesh
+    """
+    from meshmode.distributed import mpi_distribute
 
+    num_ranks = comm.Get_size()
 
-def create_parallel_grid(comm, generate_grid):
-    """Generate and distribute mesh compatibility interface."""
-    from warnings import warn
-    warn("Do not call create_parallel_grid; use generate_and_distribute_mesh "
-         "instead. This function will disappear August 1, 2021",
-         DeprecationWarning, stacklevel=2)
-    return generate_and_distribute_mesh(comm=comm, generate_mesh=generate_grid)
+    if partition_generator_func is None:
+        def partition_generator_func(mesh, tag_to_elements, num_ranks):
+            from meshmode.distributed import get_partition_by_pymetis
+            return get_partition_by_pymetis(mesh, num_ranks)
+
+    if comm.Get_rank() == 0:
```

Comment:
This all looks pretty bureaucratic. Can we do this in grudge/meshmode? What's the background to this?

Reply:
Removed from this PR (still need to figure out how to clean it up, but we'll worry about that later.)